src/Expander/Primitives.hs

```diff
@@ -512,90 +509,41 @@ typeCase t dest stx = do
 type TypePrim =
   (Kind -> SplitTypePtr -> Syntax -> Expand (), TypePatternPtr -> Syntax -> Expand ())
 
-baseType :: (forall a . TyF a) -> TypePrim
-baseType ctor = (implT, implP)
-  where
-    implT k dest stx = do
-      equateKinds stx KStar k
-      _actualName <- mustBeIdent stx
-      linkType dest ctor
-    implP dest stx = do
-      _actualName <- mustBeIdent stx
-      linkTypePattern dest
-        (TypePattern ctor)
-        []
-
-arrowType :: TypePrim
-arrowType = (implT, implP)
+typeConstructor :: TypeConstructor -> [Kind] -> TypePrim
+typeConstructor ctor argKinds = (implT, implP)
   where
     implT k dest stx = do
-      Stx _ _ oneToThreeEntries <- mustHaveEntries stx
-      case oneToThreeEntries of
-        Left (Identity _) -> do
-          equateKinds stx (kFun [KStar, KStar] KStar) k
-          linkType dest $ TyF TFun []
-        Right (Left (_, arg)) -> do
-          equateKinds stx (kFun [KStar] KStar) k
-          argDest <- scheduleType KStar arg
-          linkType dest $ TyF TFun [argDest]
-        Right (Right (_, arg, ret)) -> do
-          equateKinds stx KStar k
-          argDest <- scheduleType KStar arg
-          retDest <- scheduleType KStar ret
-          linkType dest $ tFun1 argDest retDest
+      Stx _ _ (_, args) <- mustBeCons stx
+      if length args > length argKinds
+        then throwError $ WrongTypeArity stx ctor
+                            (fromIntegral $ length argKinds)
+                            (length args)
+        else do
+          let missingArgs :: [Kind]
+              missingArgs = drop (length args) argKinds
+          equateKinds stx (kFun missingArgs KStar) k
+          argDests <- traverse (uncurry scheduleType) (zip argKinds args)
+          linkType dest $ TyF ctor argDests
     implP dest stx = do
-      Stx _ _ oneToThreeEntries <- mustHaveEntries stx
-      case oneToThreeEntries of
-        Left (Identity _) -> do
-          linkTypePattern dest
-            (TypePattern (TyF TFun []))
-            []
-        Right (Left (_, arg)) -> do
-          (sc1, n1, x1) <- prepareVar arg
-          sch <- trivialScheme tType
-          linkTypePattern dest
-            (TypePattern (TyF TFun [(n1, x1)]))
-            [(sc1, n1, x1, sch)]
-        Right (Right (_, arg, ret)) -> do
-          (sc1, n1, x1) <- prepareVar arg
-          (sc2, n2, x2) <- prepareVar ret
+      Stx _ _ (_, args) <- mustBeCons stx
+      if length args > length argKinds
+        then throwError $ WrongTypeArity stx ctor
+                            (fromIntegral $ length argKinds)
+                            (length args)
+        else do
+          varInfo <- traverse prepareVar args
           sch <- trivialScheme tType
+          -- FIXME kind check here
```

Comment:
This `FIXME` comes from `addDatatype`. Presumably this is where we will implement #196.